### PR TITLE
[SPARK-6576] [MLlib] [PySpark] DenseMatrix in PySpark should support indexing

### PR DESCRIPTION
**Pull request by MechCoder**
Support indexing in DenseMatrices in PySpark

### CODE REVIEW

**Comment by MechCoder** [comment]
cc: @davies @mengxr

File: python/pyspark/mllib/tests.py [hunk 1]
```diff
@@ -134,6 +134,13 @@ def test_sparse_vector_indexing(self):
         for ind in [4, -5, 7.8]:
             self.assertRaises(ValueError, sv.__getitem__, ind)
 
+    def test_matrix_indexing(self):
+        mat = DenseMatrix(3, 2, [0, 1, 4, 6, 8, 10])
+        for ind, val in enumerate(mat.values):
+            rowind = ind // numcols
+            colind = ind - rowind * numcols
+            self.assertEquals(mat[rowind , colind], val)
```
Comment (davies) [inline python/pyspark/mllib/tests.py]: nit: no space need before `,`

**Comment by davies** [comment]
Jenkins, OK to test.

LGTM, just one minor comments. You could run dev/lint-python to do python style check.

**Comment by MechCoder** [comment]
@davies Thanks. I fixed and amended the commit.

**Comment by MechCoder** [comment]
@davies The Python lint checks pass in my laptop, but due to some unrelated issue, it fails here, Cab you have a look?

**Comment by MechCoder** [comment]
Finally passed all test. Can this be merged now?

**Comment by davies** [comment]
@mengxr I think this one is ready to go, thanks!

**Comment by MechCoder** [comment]
@jkbradley Do you mind having a quick look at this? It is a minor PR. Thanks! :)

File: python/pyspark/mllib/linalg.py [hunk 2]
```diff
@@ -634,6 +634,17 @@ def toArray(self):
         """
         return self.values.reshape((self.numRows, self.numCols), order='F')
 
+    def __getitem__(self, indices):
+        i = indices[0]
```
Comment (mengxr) [inline python/pyspark/mllib/linalg.py]: This is cleaner:
~~~python
i, j = indices
~~~

File: python/pyspark/mllib/linalg.py [hunk 3]
```diff
@@ -634,6 +634,17 @@ def toArray(self):
...
+        j = indices[1]
+        if i > self.numRows:
```
Comment (mengxr) [inline python/pyspark/mllib/linalg.py]: `or i < 0` and please also update the error message

File: python/pyspark/mllib/linalg.py [hunk 4]
```diff
@@ -634,6 +634,17 @@ def toArray(self):
...
+            raise ValueError("Row index %d should not be greater than numRows %d." %
+                             (i, self.numRows))
+        if j > self.numCols:
```
Comment (mengxr) [inline python/pyspark/mllib/linalg.py]: `or j < 0`

File: python/pyspark/mllib/linalg.py [hunk 5]
```diff
@@ -634,6 +634,17 @@ def toArray(self):
...
+            raise ValueError("Column index %d should not be greater than numCols %d." %
+                             (i, self.numCols))
+        return self.values[i * self.numCols + j]
```
Comment (mengxr) [inline python/pyspark/mllib/linalg.py]: This is wrong. We use column-major. So it should be `i + j * self.numRows`.

File: python/pyspark/mllib/tests.py [hunk 6]
```diff
@@ -134,6 +134,14 @@ def test_sparse_vector_indexing(self):
         for ind in [4, -5, 7.8]:
             self.assertRaises(ValueError, sv.__getitem__, ind)
 
+    def test_matrix_indexing(self):
+        mat = DenseMatrix(3, 2, [0, 1, 4, 6, 8, 10])
+        numcols = mat.numCols
+        for ind, val in enumerate(mat.values):
```
Comment (mengxr) [inline python/pyspark/mllib/tests.py]: I think it should be okay to write down the checks explicitly. Or you can do this:

~~~python
expected = [[0, 6], [1, 8], [4, 10]]
for i in range(3):
  for j in range(2):
    self.assertEquals(mat[i, j], xpected[i][j])
~~~

File: python/pyspark/mllib/linalg.py (hunk 5 above)
Comment (MechCoder) [inline python/pyspark/mllib/linalg.py]: Ouch. I completely forgot that :/

**Comment by MechCoder** [comment]
@mengxr fixed!

File: python/pyspark/mllib/linalg.py [hunk 7]
```diff
@@ -634,6 +634,18 @@ def toArray(self):
         """
         return self.values.reshape((self.numRows, self.numCols), order='F')
 
+    def __getitem__(self, indices):
+        i, j = indices
+        if i > self.numRows or i < 0:
```
Comment (mengxr) [inline python/pyspark/mllib/linalg.py]: Sorry, I didn't notice this in the first pass. The bound should be `[0, numRows)`:

~~~
if i < 0 or i >= self.numRows:
  raise ValueError("Row index %d is out of range range [0, %d)" % (i, self.numRows))
~~~

Same for `j` below.

Comment (MechCoder) [inline python/pyspark/mllib/linalg.py]: Extremely sorry for these careless mistakes. I was in a real hurry.

**Comment by MechCoder** [comment]
@mengxr fixed !

**Comment by mengxr** [comment]
LGTM. Merged into master. Thanks!

**Pull request closed without merging**